### DISCUSSION TITLE
externalized dependencies

### DISCUSSION THREAD

**Discussion by erwinheldy** [privatenumber/pkgroll#42] [General]
Why does the build output include syntax for importing packages in dependencies, whereas in the readme it says if packages in dependencies will be externalized

![image](https://github.com/privatenumber/pkgroll/assets/1826289/3023ba73-6681-4b0e-8e7b-c73a7ae0f876)

**Comment by privatenumber** [comment]
Perhaps you're missing what externalization does?

If not, what is your expected behavior?

Here are some docs on it:
- https://rollupjs.org/configuration-options/#external
- https://esbuild.github.io/api/#external
- https://webpack.js.org/configuration/externals/#root

ChatGPT explanation:
https://chat.openai.com/share/ad4981d9-854d-444b-b231-b2cf86dbc57c

**Reply by erwinheldy** [reply]
I think for the pkroll build result, the line:
```javascript
import 'node:http';
...
```
not needed.

I also tried the bundle using esbuild, the result of the bundle that I got was no syntax for importing the dependencies since i included it as `external`.

**Reply by privatenumber** [reply]
Why's it not needed?

Impossible for me to determine given I don't know your input.

**Reply by erwinheldy** [reply]
Here's a simple example, the left side is the source while the right side is the build result.

![image](https://github.com/privatenumber/pkgroll/assets/1826289/66684ae3-8dab-4696-8f9d-69e92d821116)

Take a look at the `cli.js` file, line 4: `import 'node:fs/promises';` I don't think it's needed because it's already imported in the `index.js` file, what do you think?

**Reply by erwinheldy** [reply]
For comparison, here is the result of using esbuild with the command:
```bash
esbuild src/index.ts --format=esm --platform=node --outdir=dist --bundle
esbuild src/cli.ts --format=esm --platform=node --outdir=dist
```
![image](https://github.com/privatenumber/pkgroll/assets/1826289/da064019-efed-42e1-894f-0627b13b16b5)

**Reply by privatenumber** [reply]
So JSYK the externalization logic is handled by Rollup, rather than pkgroll: [REPL](https://rollupjs.org/repl/?version=3.27.2&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMHJlYWRGaWxlJTIwJTdEJTIwZnJvbSUyMCdub2RlJTNBZnMlMkZwcm9taXNlcyclM0IlNUNuZXhwb3J0JTIwY29uc3QlMjByZWFkJTIwJTNEJTIwcmVhZEZpbGUlM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTdEJTJDJTdCJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMHByb2Nlc3MlMjBmcm9tJTIwJ25vZGUlM0Fwcm9jZXNzJyUzQiU1Q25pbXBvcnQlMjAlN0IlMjByZWFkJTIwJTdEJTIwZnJvbSUyMCcuJTJGbWFpbi5qcyclM0IlNUNuJTVDbmNvbnNvbGUubG9nKHJlYWQocHJvY2VzcykpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIyY2xpLmpzJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMm91dHB1dCUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTdEJTJDJTIydHJlZXNoYWtlJTIyJTNBJTIyc21hbGxlc3QlMjIlN0QlN0Q=)

The reason the imports are moved up is explained here:
https://rollupjs.org/faqs/#why-do-additional-imports-turn-up-in-my-entry-chunks-when-code-splitting

TLDR: it's a performance optimization so the JS engine can resolve dependencies faster

**Reply by erwinheldy** [reply]
Great, thanks for the explanation.
